File: kubernetes-informer/src/test/java/io/micronaut/kubernetes/informer/ConfigMapInformer.java
/*
 * Copyright 2021 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.micronaut.kubernetes.informer;

import io.kubernetes.client.informer.ResourceEventHandler;
import io.kubernetes.client.openapi.models.V1ConfigMap;
import io.kubernetes.client.openapi.models.V1ConfigMapList;
import io.micronaut.context.annotation.Requires;
import jakarta.inject.Singleton;
import org.apache.commons.compress.utils.Lists;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;

//tag::listener[]
@Requires(property = "spec.name", value = "ConfigMapInformerSpec")
@Singleton
@Informer(apiType = V1ConfigMap.class, apiListType = V1ConfigMapList.class, resourcePlural = "configmaps") // <1>
public class ConfigMapInformer implements ResourceEventHandler<V1ConfigMap> { // <2>

    //end::listener[]
    public static final Logger LOG = LoggerFactory.getLogger(ConfigMapInformer.class);

    private final List<V1ConfigMap> added = Lists.newArrayList();
    private final List<V1ConfigMap> updated = Lists.newArrayList();
    private final List<V1ConfigMap> deleted = Lists.newArrayList();

    public List<V1ConfigMap> getAdded() {
        return added;
    }

    public List<V1ConfigMap> getUpdated() {
        return updated;
    }

    public List<V1ConfigMap> getDeleted() {
        return deleted;
    }

    //tag::listener[]
    @Override
    public void onAdd(V1ConfigMap obj) {
        //end::listener[]
        LOG.info("ADDED CONFIG MAP: {}", obj);
        added.add(obj);
        //tag::listener[]
    }


    @Override
    public void onUpdate(V1ConfigMap oldObj, V1ConfigMap newObj) {
        //end::listener[]
        LOG.info("UPDATED CONFIG MAP: {}", newObj);
        updated.add(newObj);
        //tag::listener[]
    }

    @Override
    public void onDelete(V1ConfigMap obj, boolean deletedFinalStateUnknown) {
        //end::listener[]
        LOG.info("DELETED CONFIG MAP: {}", obj);
        deleted.add(obj);
        //tag::listener[]
    }
}
//end::listener[]

File: examples/example-kubernetes-client-openapi-informer-python/src/main/python/micronaut/client/configmap/ConfigMapInformer.py
# tag::class[]
import logging

from jakarta.annotation import PostConstruct
from micronaut.context.annotation import Context
from micronaut.kubernetes.client.openapi.informer import SharedIndexInformerFactory
from micronaut.kubernetes.client.openapi.informer.handler import ResourceEventHandler
from micronaut.kubernetes.client.openapi.model import V1ConfigMap
from micronaut.kubernetes.client.openapi.resolver import NamespaceResolver
# end::class[]
from micronaut.context.annotation import Requires
from micronaut.context.env import Environment
# tag::class[]

LOG = logging.getLogger(__name__)

# end::class[]

@Requires(env=Environment.KUBERNETES)
# tag::class[]
@Context
class ConfigMapInformer:

    def __init__(self, shared_index_informer_factory: SharedIndexInformerFactory, namespace_resolver: NamespaceResolver):
        self.shared_index_informer_factory = shared_index_informer_factory
        self.namespace_resolver = namespace_resolver

    @PostConstruct
    def initialize(self) -> None:
        namespace = self.namespace_resolver.resolveNamespace()
        shared_index_informer = self.shared_index_informer_factory.sharedIndexInformerFor(
            V1ConfigMap, namespace)
        shared_index_informer.addEventHandler(self.ConfigMapEventHandler())

    class ConfigMapEventHandler(ResourceEventHandler[V1ConfigMap]):

        def onAdd(self, obj: V1ConfigMap) -> None:
            LOG.info("%s config map added!", obj.getMetadata().getName())

        def onUpdate(self, old_obj: V1ConfigMap, new_obj: V1ConfigMap) -> None:
            LOG.info("%s config map updated!", old_obj.getMetadata().getName())

        def onDelete(self, obj: V1ConfigMap, deleted_final_state_unknown: bool) -> None:
            LOG.info("%s config map deleted!", obj.getMetadata().getName())
# end::class[]
